dot/rpc/subscription/websocket.go

```diff
@@ -161,3 +163,206 @@ func (c *WSConn) HandleComm() {
+		return 0, fmt.Errorf("error StorageAPI not set")
+	}
+
+	myObs := &StorageObserver{
+		filter: make(map[string][]byte),
+		wsconn: c,
+	}
+
+	pA, ok := params.([]interface{})
+	if !ok {
+		return 0, fmt.Errorf("unknown parameter type")
+	}
+	for _, param := range pA {
+		switch p := param.(type) {
+		case []interface{}:
+			for _, pp := range param.([]interface{}) {
+				data, ok := pp.(string)
+				if !ok {
+					return 0, fmt.Errorf("unknown parameter type")
+				}
+				myObs.filter[data] = []byte{}
+			}
+		case string:
+			myObs.filter[p] = []byte{}
+		default:
+			return 0, fmt.Errorf("unknown parameter type")
+		}
+	}
+
+	c.qtyListeners++
+	myObs.id = c.qtyListeners
+
+	c.StorageAPI.RegisterStorageObserver(myObs)
+
+	c.Subscriptions[myObs.id] = myObs
+
+	initRes := newSubscriptionResponseJSON(myObs.id, reqID)
+	c.safeSend(initRes)
+
+	return myObs.id, nil
+}
+
+func (c *WSConn) initBlockListener(reqID float64) (int, error) {
+	bl := &BlockListener{
+		Channel: make(chan *types.Block),
+		wsconn:  c,
+	}
+
+	if c.BlockAPI == nil {
+		c.safeSendError(reqID, nil, "error BlockAPI not set")
+		return 0, fmt.Errorf("error BlockAPI not set")
+	}
+	chanID, err := c.BlockAPI.RegisterImportedChannel(bl.Channel)
+	if err != nil {
+		return 0, err
+	}
+	bl.ChanID = chanID
+	c.qtyListeners++
+	bl.subID = c.qtyListeners
+	c.Subscriptions[bl.subID] = bl
+	c.BlockSubChannels[bl.subID] = chanID
+	initRes := newSubscriptionResponseJSON(bl.subID, reqID)
+	c.safeSend(initRes)
+
+	return bl.subID, nil
+}
+
+func (c *WSConn) initBlockFinalizedListener(reqID float64) (int, error) {
+	bfl := &BlockFinalizedListener{
+		channel: make(chan *types.Header),
+		wsconn:  c,
+	}
+
+	if c.BlockAPI == nil {
+		c.safeSendError(reqID, nil, "error BlockAPI not set")
+		return 0, fmt.Errorf("error BlockAPI not set")
+	}
+	chanID, err := c.BlockAPI.RegisterFinalizedChannel(bfl.channel)
+	if err != nil {
+		return 0, err
+	}
+	bfl.chanID = chanID
+	c.qtyListeners++
+	bfl.subID = c.qtyListeners
+	c.Subscriptions[bfl.subID] = bfl
+	c.BlockSubChannels[bfl.subID] = chanID
+	initRes := newSubscriptionResponseJSON(bfl.subID, reqID)
+	c.safeSend(initRes)
+
+	return bfl.subID, nil
+}
+
+func (c *WSConn) initExtrinsicWatch(reqID float64, params interface{}) (int, error) {
+	pA := params.([]interface{})
+	extBytes, err := common.HexToBytes(pA[0].(string))
+	if err != nil {
+		return 0, err
+	}
+
+	// listen for built blocks
+	esl := &ExtrinsicSubmitListener{
+		importedChan:  make(chan *types.Block),
+		wsconn:        c,
+		extrinsic:     types.Extrinsic(extBytes),
+		finalizedChan: make(chan *types.Header),
+	}
+
+	if c.BlockAPI == nil {
+		return 0, fmt.Errorf("error BlockAPI not set")
+	}
+	esl.importedChanID, err = c.BlockAPI.RegisterImportedChannel(esl.importedChan)
+	if err != nil {
+		return 0, err
+	}
+
+	esl.finalizedChanID, err = c.BlockAPI.RegisterFinalizedChannel(esl.finalizedChan)
+	if err != nil {
+		return 0, err
+	}
+
```

Comment:
just curious, done submitAndWatchExtrinsic notify both on the block being built as well as finalized?

Reply:
Yes, it has a "inBlock" status for when the extrinsic has been included in block, and "finalized" status for when the extrinsic has been finalized by GRANDPA.